File: python/5programe.py

class Nod:
    def __init__(self,valoare,index):
        self.valoare = valoare
        self.index = index
        self.dreapta = None
        self.stanga = None

class Aux:
    def __init__(self,valoare,index,parinte,pozitie):
        self.valoare = valoare
        self.index = index
        self.parinte = parinte
        self.pozitie = pozitie

class ArboreBinar:
    def __init__(self):
        self.radacina = None
    
    def parcurgere(self):
        queue = []
        queue.append(self.radacina)
        while len(queue)>0:
            currentNod = queue.pop(0)
            print(str(currentNod.index) + " ")
            if currentNod.stanga is not None:
                queue.append(currentNod.stanga)
            if currentNod.dreapta is not None:
                queue.append(currentNod.dreapta)
        

class FileParser:
    FILE_PATH = "/home/school/Projects/5programe/"
    def readLines(self):
        radacina = False
        set = []
        fileName = raw_input('Dati numele fisierului: ')
        numeFisier = self.FILE_PATH+fileName
        filer = open(numeFisier, 'r')
        Lines = filer.readlines()
        for line in Lines:
            print(line)
            dateList = line.split(',')
            print(dateList)
            if len(dateList) != 4:
                raise Exception("Datele sunt insuficiente")
            aux = Aux(dateList[3],int(dateList[0]),int(dateList[1]),int(dateList[2]))
            if aux.parinte < 0:
                if radacina is False:
                    print("are")
                    radacina = True
                else:
                    raise Exception("Arborele nu poate avea 2 radacini")
            set.append(aux)
        filer.close()
        if radacina == False:
            raise Exception("Arborele trebuie sa aiba radacina")
        return set

    
    def listToNodeHashap(self,list):
        map = {}
        for x in list:
            nod = Nod(x.valoare, x.index)
            map[x.index] = nod
        return map
    
    def genereazaArbore(self):
        list = self.readLines()
        arbore = ArboreBinar()
        map = self.listToNodeHashap(list)
        for item in list:
            if item.parinte < 0:
                arbore.radacina = map[item.index]
            else:
                nod = map[item.index]
                parinte = map[item.parinte]
                if item.pozitie == 0:
                    parinte.stanga = nod
                else:
                    if item.pozitie == 1:
                        parinte.dreapta = nod
                    else:
                        raise Exception("Parintele contine nodul")
        return arbore           
        
            

fileparser = FileParser()
arbore = fileparser.genereazaArbore()
arbore.parcurgere()

File: php/5programe.php
<?php
class Aux{
    public $index;
    public $valoare;
    public $parinte;
    public $pozitie;

    function __construct($v,$i,$p,$poz){
        $this->valoare = $v;
        $this->index = $i;
        $this->parinte = $p;
        $this->pozitie = $poz;
    }
}
class Nod{
    public $stanga;
    public $dreapta;
    public $index;
    public $valoare;

    function __construct($valoare, $index){
        $this->valoare = $valoare;
        $this->index = $index;
    }
}
class FileParser{
    public const FILE_PATH = "/home/school/Projects/5programe/";

    function readLines(){
        $fisier = readline('Enter a string: ');
        print(self::FILE_PATH.$fisier);
        $file = fopen(self::FILE_PATH.$fisier,"r");
        $set = array();
        $radacina = false;
        if($file){
            while (($line = fgets($file)) !== false) {
                $dateList = explode(",",$line);
                if(sizeof($dateList)!=4){
                    throw new Exception("Nodul are date insuficiente");
                }
                $aux = new Aux($dateList[3],$dateList[0],$dateList[1],$dateList[2]);
                if($aux->parinte < 0){
                    if(!$radacina){
                        $radacina = true;
                    }else{
                        throw new Exception("Arborele nu poate avea 2 radacini");
                    }
                }
                array_push($set,$aux);

            }
            fclose($file);
            if(!$radacina){
                throw new Exception("Arborele trebuie sa aiba o radacina");
            }
            return $set;
        }

    }

    function listNodeToHashmap($list){
        $map = array();
        foreach($list as $item){
            $nod = new Nod($item->valoare,$item->index);
            $map[$item->index] = $nod;
        }
        return $map;
    }

    function genereazaArbore(){
        $list = $this->readLines();
        $arbore = new ArboreBinar();
        $map = $this->listNodeToHashmap($list);
        foreach($list as $item){
            if($item->parinte < 0){
                $arbore->radacina = $map[$item->index];

            }else{
                $nod = $map[$item->index];
                $parinte = $map[$item->parinte];
                if($item->pozitie == 0){
                    $parinte->stanga = $nod;
                }else {
                    if ($item->pozitie == 1) {
                        $parinte->dreapta = $nod;
                    } else {
                            throw new Exception("Parintele contine nodul");
                    }
                }
            }
        }
        return $arbore;
    }
}
class ArboreBinar{
    public $radacina;

    function parcurgere(){
        $queue = new SplQueue();
        $queue->enqueue($this->radacina);
        while(sizeof($queue)>0) {
            $currentNode = $queue->dequeue();
            print($currentNode->index." ");
            if ($currentNode->stanga != null)
                $queue->enqueue($currentNode->stanga);
            if ($currentNode->dreapta != null)
                $queue->enqueue($currentNode->dreapta);
        }
    }
}

$fileParser = new FileParser();
$arbore = $fileParser->genereazaArbore();
$arbore->parcurgere();

?>